returnn/tf/layers/basic.py

```diff
@@ -913,37 +913,204 @@ def transform_config_dict(cls, d, network, get_layer):
+    common_axes_input, common_axes_position = zip(*common_axes_pairs) if common_axes_pairs else ([], [])
+    common_axes_input, common_axes_position = list(common_axes_input), list(common_axes_position)
+    specific_input_axes = [
+      axis for axis in range(input_data.batch_ndim) if axis not in common_axes_input and not axis == old_gather_axis]
+    specific_position_axes = [axis for axis in range(position_data.batch_ndim) if axis not in common_axes_position]
+    return common_axes_input, common_axes_position, specific_input_axes, specific_position_axes
+
+  @classmethod
+  def translate_input_axis(
+    cls, input_axis, old_gather_axis, common_axes_input, input_axes, position_axes):
+    """
+    :param int input_axis: batch axis of input_data
+    :param int old_gather_axis: gather axis of ``input_data`` counted with batch dim
+    :param list[int] common_axes_input:
+    :param list[int] input_axes:
+    :param list[int] position_axes:
+    :rtype: int|None
+    :return: batch axis of output
+    """
+    if input_axis in common_axes_input:
+      return common_axes_input.index(input_axis)
+    elif input_axis == old_gather_axis:
+      return None  # the gather axis will not be present in the output
+    else:
+      return len(common_axes_input) + len(position_axes) + input_axes.index(input_axis)
 
   @classmethod
-  def get_out_data_from_opts(cls, name, sources, axis, **kwargs):
+  def translate_position_axis(
+    cls, position_axis, common_axes_input, common_axes_position, position_axes):
+    """
+    :param int position_axis: batch axis of position_data
+    :param list[int] common_axes_input:
+    :param list[int] common_axes_position:
+    :param list[int] position_axes:
+    :rtype: int
+    :return: batch axis of output
+    """
+    if position_axis in common_axes_position:
+      return common_axes_position.index(position_axis)
+    else:
+      return len(common_axes_input) + position_axes.index(position_axis)
+
+  def get_dep_layers(self):
+    """
+    :rtype: list[LayerBase]
+    """
+    if isinstance(self.position, LayerBase):
+      return super(GatherLayer, self).get_dep_layers() + [self.position]
+    else:
+      return super(GatherLayer, self).get_dep_layers()
+
+  @classmethod
+  def get_out_data_from_opts(cls, name, sources, position, axis, **kwargs):
     """
     :param str name:
     :param list[LayerBase] sources:
+    :param LayerBase|int position:
     :param str axis:
     :rtype: Data
     """
-    out = get_concat_sources_data_template(sources, name="%s_output" % name)
-    axis = out.get_axis_from_description(axis, allow_int=False)
-    return out.copy_template_excluding_axis(axis)
+    input_data = get_concat_sources_data_template(sources)
+    gather_axis = input_data.get_axis_from_description(axis, allow_int=False)
+
+    if isinstance(position, int):
+      position = ConstantLayer.get_out_data_from_opts(
+        name="constant_position", network=kwargs["network"], value=position, dtype="int32")
```

Comment:
You don't need to do that so complicated.
```suggestion
      position = Data(name="constant_position", dtype="int32", shape=())
```